File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/wallattachments/WallSearchCommentsAttachmentsPresenter.java
package dev.ragnarok.fenrir.mvp.presenter.wallattachments;

import static dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime;
import static dev.ragnarok.fenrir.util.Utils.safeCountOf;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.Injection;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.domain.ICommentsInteractor;
import dev.ragnarok.fenrir.domain.Repository;
import dev.ragnarok.fenrir.domain.impl.CommentsInteractor;
import dev.ragnarok.fenrir.model.Comment;
import dev.ragnarok.fenrir.model.CommentedType;
import dev.ragnarok.fenrir.mvp.presenter.base.PlaceSupportPresenter;
import dev.ragnarok.fenrir.mvp.view.wallattachments.IWallSearchCommentsAttachmentsView;
import dev.ragnarok.fenrir.util.RxUtils;
import io.reactivex.rxjava3.disposables.CompositeDisposable;

public class WallSearchCommentsAttachmentsPresenter extends PlaceSupportPresenter<IWallSearchCommentsAttachmentsView> {

    private final List<Comment> data;
    private final List<Integer> posts;
    private final ICommentsInteractor interactor;
    private final int owner_id;
    private final CompositeDisposable actualDataDisposable = new CompositeDisposable();
    private int loaded;
    private int offset;
    private int index;
    private boolean actualDataReceived;
    private boolean endOfContent;
    private boolean actualDataLoading;

    public WallSearchCommentsAttachmentsPresenter(int accountId, int ownerId, @NonNull List<Integer> posts, @Nullable Bundle savedInstanceState) {
        super(accountId, savedInstanceState);
        this.posts = posts;
        owner_id = ownerId;
        data = new ArrayList<>();
        interactor = new CommentsInteractor(Injection.provideNetworkInterfaces(), Injection.provideStores(), Repository.INSTANCE.getOwners());
        loadActualData();
    }

    @Override
    public void onGuiCreated(@NonNull IWallSearchCommentsAttachmentsView view) {
        super.onGuiCreated(view);
        view.displayData(data);

        resolveToolbar();
    }

    private void loadActualData() {
        actualDataLoading = true;

        resolveRefreshingView();

        int accountId = getAccountId();
        actualDataDisposable.add(interactor.getCommentsNoCache(accountId, owner_id, posts.get(index), offset)
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(this::onActualDataReceived, this::onActualDataGetError));

    }

    private void onActualDataGetError(Throwable t) {
        actualDataLoading = false;
        callView(v -> showError(v, getCauseIfRuntime(t)));

        resolveRefreshingView();
    }

    private void onActualDataReceived(List<Comment> res) {
        offset += 100;
        actualDataLoading = false;
        endOfContent = res.isEmpty() && index == posts.size() - 1;
        if (res.isEmpty()) {
            index++;
            offset = 0;
        }
        actualDataReceived = true;
        if (endOfContent)
            callResumedView(v -> v.onSetLoadingStatus(2));

        int startSize = data.size();
        loaded += res.size();
        data.addAll(res);
        resolveToolbar();
        callView(view -> view.notifyDataAdded(startSize, res.size() - startSize));
        resolveRefreshingView();
    }

    @Override
    public void onGuiResumed() {
        super.onGuiResumed();
        resolveRefreshingView();
    }

    private void resolveRefreshingView() {
        callResumedView(v -> v.showRefreshing(actualDataLoading));
        if (!endOfContent)
            callResumedView(v -> v.onSetLoadingStatus(actualDataLoading ? 1 : 0));
    }

    private void resolveToolbar() {
        callView(v -> {
            v.setToolbarTitle(getString(R.string.attachments_in_wall));
            v.setToolbarSubtitle(getString(R.string.comments, safeCountOf(data)) + " " + getString(R.string.comments_analized, loaded));
        });
    }

    @Override
    public void onDestroyed() {
        actualDataDisposable.dispose();
        super.onDestroyed();
    }

    public boolean fireScrollToEnd() {
        if (!endOfContent && actualDataReceived && !actualDataLoading) {
            loadActualData();
            return false;
        }
        return true;
    }

    public void fireRefresh() {

        actualDataDisposable.clear();
        actualDataLoading = false;

        offset = 0;
        index = 0;
        loaded = 0;
        data.clear();
        callView(IWallSearchCommentsAttachmentsView::notifyDataSetChanged);
        loadActualData();
    }

    public void fireReplyToOwnerClick(int commentId) {
        for (int y = 0; y < data.size(); y++) {
            Comment comment = data.get(y);
            if (comment.getId() == commentId) {
                comment.setAnimationNow(true);

                int finalY = y;
                callView(v -> v.notifyItemChanged(finalY));
                callView(v -> v.moveFocusTo(finalY));
                return;
            } else if (comment.hasThreads()) {
                for (int s = 0; s < comment.getThreads().size(); s++) {
                    Comment thread = comment.getThreads().get(s);
                    if (thread.getId() == commentId) {
                        thread.setAnimationNow(true);
                        int finalY = y;
                        callView(v -> v.notifyItemChanged(finalY));
                        callView(v -> v.moveFocusTo(finalY));
                        return;
                    }
                }
            }
        }
    }

    private String getApiCommentType(Comment comment) {
        switch (comment.getCommented().getSourceType()) {
            case CommentedType.PHOTO:
                return "photo_comment";
            case CommentedType.POST:
                return "comment";
            case CommentedType.VIDEO:
                return "video_comment";
            case CommentedType.TOPIC:
                return "topic_comment";
            default:
                throw new IllegalArgumentException();
        }
    }

    public void fireWhoLikesClick(Comment comment) {
        callView(v -> v.goToLikes(getAccountId(), getApiCommentType(comment), owner_id, comment.getId()));
    }

    public void fireGoCommentPostClick(Comment comment) {
        callView(v -> v.goToPost(getAccountId(), comment.getCommented().getSourceOwnerId(), comment.getCommented().getSourceId()));
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/wall/wallattachments/wallsearchcommentsattachments/WallSearchCommentsAttachmentsPresenter.kt
package dev.ragnarok.fenrir.fragment.wall.wallattachments.wallsearchcommentsattachments

import android.os.Bundle
import dev.ragnarok.fenrir.Includes.networkInterfaces
import dev.ragnarok.fenrir.Includes.stores
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.domain.ICommentsInteractor
import dev.ragnarok.fenrir.domain.Repository.owners
import dev.ragnarok.fenrir.domain.impl.CommentsInteractor
import dev.ragnarok.fenrir.fragment.base.PlaceSupportPresenter
import dev.ragnarok.fenrir.fromIOToMain
import dev.ragnarok.fenrir.model.Comment
import dev.ragnarok.fenrir.model.CommentedType
import dev.ragnarok.fenrir.util.Utils.getCauseIfRuntime
import dev.ragnarok.fenrir.util.Utils.safeCountOf
import io.reactivex.rxjava3.disposables.CompositeDisposable

class WallSearchCommentsAttachmentsPresenter(
    accountId: Long,
    private val owner_id: Long,
    private val posts: List<Int>,
    savedInstanceState: Bundle?
) : PlaceSupportPresenter<IWallSearchCommentsAttachmentsView>(accountId, savedInstanceState) {
    private val data: MutableList<Comment>
    private val interactor: ICommentsInteractor
    private val actualDataDisposable = CompositeDisposable()
    private var loaded = 0
    private var offset = 0
    private var index = 0
    private var actualDataReceived = false
    private var endOfContent = false
    private var actualDataLoading = false
    override fun onGuiCreated(viewHost: IWallSearchCommentsAttachmentsView) {
        super.onGuiCreated(viewHost)
        viewHost.displayData(data)
        resolveToolbar()
    }

    private fun loadActualData() {
        actualDataLoading = true
        resolveRefreshingView()
        actualDataDisposable.add(interactor.getCommentsNoCache(
            accountId,
            owner_id,
            posts[index],
            offset
        )
            .fromIOToMain()
            .subscribe({ onActualDataReceived(it) }) {
                onActualDataGetError(
                    it
                )
            })
    }

    private fun onActualDataGetError(t: Throwable) {
        actualDataLoading = false
        showError(getCauseIfRuntime(t))
        resolveRefreshingView()
    }

    private fun onActualDataReceived(res: List<Comment>) {
        offset += 100
        actualDataLoading = false
        endOfContent = res.isEmpty() && index == posts.size - 1
        if (res.isEmpty()) {
            index++
            offset = 0
        }
        actualDataReceived = true
        if (endOfContent) resumedView?.onSetLoadingStatus(
            2
        )
        val startSize = data.size
        loaded += res.size
        data.addAll(res)
        resolveToolbar()
        view?.notifyDataAdded(
            startSize,
            res.size - startSize
        )
        resolveRefreshingView()
    }

    public override fun onGuiResumed() {
        super.onGuiResumed()
        resolveRefreshingView()
    }

    private fun resolveRefreshingView() {
        resumedView?.showRefreshing(
            actualDataLoading
        )
        if (!endOfContent) resumedView?.onSetLoadingStatus(
            if (actualDataLoading) 1 else 0
        )
    }

    private fun resolveToolbar() {
        view?.let {
            it.toolbarTitle(getString(R.string.attachments_in_wall))
            it.toolbarSubtitle(
                getString(
                    R.string.comments,
                    safeCountOf(data)
                ) + " " + getString(R.string.comments_analized, loaded)
            )
        }
    }

    override fun onDestroyed() {
        actualDataDisposable.dispose()
        super.onDestroyed()
    }

    fun fireScrollToEnd(button: Boolean): Boolean {
        if (!endOfContent && actualDataReceived && !actualDataLoading) {
            loadActualData()
            return false
        }
        if (button && endOfContent) resumedView?.onSetLoadingStatus(2)
        return true
    }

    fun fireRefresh() {
        actualDataDisposable.clear()
        actualDataLoading = false
        offset = 0
        index = 0
        loaded = 0
        data.clear()
        view?.notifyDataSetChanged()
        loadActualData()
    }

    fun fireReplyToOwnerClick(commentId: Int) {
        for (y in data.indices) {
            val comment = data[y]
            if (comment.getObjectId() == commentId) {
                comment.setAnimationNow(true)
                view?.notifyItemChanged(
                    y
                )
                view?.moveFocusTo(y)
                return
            } else if (comment.hasThreads()) {
                for (s in comment.threads.orEmpty()) {
                    if (s.getObjectId() == commentId) {
                        s.setAnimationNow(true)
                        view?.notifyItemChanged(
                            y
                        )
                        view?.moveFocusTo(y)
                        return
                    }
                }
            }
        }
    }

    private fun getApiCommentType(comment: Comment): String {
        return when (comment.commented.sourceType) {
            CommentedType.PHOTO -> "photo_comment"
            CommentedType.POST -> "comment"
            CommentedType.VIDEO -> "video_comment"
            CommentedType.TOPIC -> "topic_comment"
            else -> throw IllegalArgumentException()
        }
    }

    fun fireWhoLikesClick(comment: Comment) {
        view?.goToLikes(
            accountId,
            getApiCommentType(comment),
            owner_id,
            comment.getObjectId()
        )
    }

    fun fireGoCommentPostClick(comment: Comment) {
        view?.goToPost(
            accountId,
            comment.commented.sourceOwnerId,
            comment.commented.sourceId
        )
    }

    init {
        data = ArrayList()
        interactor = CommentsInteractor(networkInterfaces, stores, owners)
        loadActualData()
    }
}